### DISCUSSION TITLE
Vits with d-vector

### DISCUSSION THREAD

**Discussion by KyuubiYoru** [coqui-ai/TTS#1993] [General Q&A]
I tried to use Vits with d-vector, but I got this error. 
I used https://github.com/coqui-ai/TTS/releases/download/speaker_encoder_model/model_se.pth.tar to compute the embeddings. All of the audio files in my dataset are mono.

> > TRAINING (2022-09-19 18:28:54) 
> /home/kyuubi/.local/lib/python3.10/site-packages/torch/functional.py:606: UserWarning: stft will soon require the return_complex parameter be given for real inputs, and will further require that return_complex=True in a future PyTorch release. (Triggered internally at  ../aten/src/ATen/native/SpectralOps.cpp:800.)
>   return _VF.stft(input, n_fft, hop_length, win_length, window,  # type: ignore[attr-defined]
>  ! Run is kept in /home/kyuubi/TTS/recipes/vctk/vits/5_vctk_lj_b60-September-19-2022_06+28PM-0a112f78
> Traceback (most recent call last):
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1533, in fit
>     self._fit()
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1517, in _fit
>     self.train_epoch()
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1282, in train_epoch
>     _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1135, in train_step
>     outputs, loss_dict_new, step_time = self._optimize(
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 996, in _optimize
>     outputs, loss_dict = self._model_train_step(batch, model, criterion, optimizer_idx=optimizer_idx)
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 954, in _model_train_step
>     return model.train_step(*input_args)
>   File "/home/kyuubi/TTS/TTS/tts/models/vits.py", line 1246, in train_step
>     outputs = self.forward(
>   File "/home/kyuubi/TTS/TTS/tts/models/vits.py", line 1028, in forward
>     o = self.waveform_decoder(z_slice, g=g)
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/home/kyuubi/TTS/TTS/vocoder/models/hifigan_generator.py", line 250, in forward
>     o = o + self.cond_layer(g)
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1130, in _call_impl
>     return forward_call(*input, **kwargs)
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 307, in forward
>     return self._conv_forward(input, self.weight, self.bias)
>   File "/home/kyuubi/.local/lib/python3.10/site-packages/torch/nn/modules/conv.py", line 303, in _conv_forward
>     return F.conv1d(input, weight, bias, self.stride,
> TypeError: conv1d() received an invalid combination of arguments - got (NoneType, Parameter, Parameter, tuple, tuple, tuple, int), but expected one of:
>  * (Tensor input, Tensor weight, Tensor bias, tuple of ints stride, tuple of ints padding, tuple of ints dilation, int groups)
>       didn't match because some of the arguments have invalid types: (NoneType, Parameter, Parameter, tuple, tuple, tuple, int)
>  * (Tensor input, Tensor weight, Tensor bias, tuple of ints stride, str padding, tuple of ints dilation, int groups)
>       didn't match because some of the arguments have invalid types: (NoneType, Parameter, Parameter, tuple, tuple, tuple, int)

**Comment by harmlessman** [comment]
see answer of https://github.com/coqui-ai/TTS/discussions/1939

I've been through the same thing.
If you set d_vector in config.json and proceed with training, there will be no error.
I don't know the cause of the error.

**Reply by KyuubiYoru** [reply]
Seems like i have a second problem, now i get,

>  > TRAINING (2022-09-21 04:51:32) 
 ! Run is kept in /home/kyuubi/TTS/recipes/vctk/vits/8_vctk_lj_b60-September-21-2022_04+51AM-0a112f78
Traceback (most recent call last):
  File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1533, in fit
    self._fit()
  File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1517, in _fit
    self.train_epoch()
  File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1282, in train_epoch
    _, _ = self.train_step(batch, batch_num_steps, cur_step, loader_start_time)
  File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 1099, in train_step
    batch = self.format_batch(batch)
  File "/home/kyuubi/.local/lib/python3.10/site-packages/trainer/trainer.py", line 896, in format_batch
    batch = self.model.format_batch(batch)
  File "/home/kyuubi/TTS/TTS/tts/models/vits.py", line 1473, in format_batch
    d_vectors = [d_vector_mapping[w]["embedding"] for w in batch["audio_files"]]
  File "/home/kyuubi/TTS/TTS/tts/models/vits.py", line 1473, in <listcomp>
    d_vectors = [d_vector_mapping[w]["embedding"] for w in batch["audio_files"]]
KeyError: 'LJ013-0177.wav'

**Reply by KyuubiYoru** [reply]
Update to the news version now it works with your trick.

Second problem was fixed with https://github.com/coqui-ai/TTS/commit/3faccbda978df0ccbb8d28820545c1989dfc16d2
